### PR TITLE
[Model] Jamba support

### LINKED ISSUE #3690: [New Model]: Jamba (MoE Mamba from AI21)
### The model to consider.

https://huggingface.co/ai21labs/Jamba-v0.1

### The closest model vllm already supports.

_No response_

### What's your difficulty of supporting the model you want?

Jamba is a new model and interesting hybrid architecture MoE Mamba model that look very promising. 

https://www.ai21.com/blog/announcing-jamba

### PR DESCRIPTION
**Pull request by mzusman**
Add Jamba support to vLLM, 
This PR comprises three parts: 
1. The Jamba modeling file that encapsulates the Jamba model weights and logic itself and the mamba cache management.
2. Passage of the requests ids of the sequence groups and their sequence ids into the modeling file in order to be able to manage the cache.
3. Passage of the finished request ids into the modeling file as well in order the clean the allocated cache on finished requests

FIX #3690

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Cool!

**Comment by WoosukKwon** [comment]
@zhuohan123 Are you still on this PR?

**Comment by zhuohan123** [comment]
@WoosukKwon Yes I am. Still looking!

File: vllm/worker/model_runner.py
Comment (zhuohan123) [inline vllm/worker/model_runner.py:969]: In vLLM, we typically put the logic of figuring out the memory address for each request in the driver instead of worker/model runner. For Transformer-based models, the logic is implemented in `BlockSpaceManager` class. I believe we should do the same thing for mamba as well. We can create another `FixedSpaceManager` class to manage the memory for mamba-style models.

File: vllm/worker/model_runner.py
Comment (zhuohan123) [inline vllm/worker/model_runner.py:982]: If we perform indexing here, why do we still need the following copy on line 964-957?

```
                        self.mamba_cache[0][:, f_free_index].copy_(
                            self.mamba_cache[0][:, i_exist])
                        self.mamba_cache[1][:, f_free_index].copy_(
                            self.mamba_cache[1][:, i_exist])
```

**Review by zhuohan123 — COMMENTED** [review]
Thanks for the work! Sorry for the delay. My two main comments on the new memory management part for mamba:

- In vLLM, the high-level memory management is typically done in the scheduler/driver. Right now, this PR does the memory management in the worker/model_runner. I think it's a better idea to follow the previous style.
- I think the memory manager can be more general and should not limited to mamba. Right now, the code is a bit too mamba-specific (e.g., `conv_state` and `ssm_state`). When outside the model code, vLLM should not need to know what exactly the state means for each model. This can help us better implement other models like recurrent gemma.

File: vllm/worker/model_runner.py
Comment (mzusman) [reply vllm/worker/model_runner.py:883]: Lines 964-957 are used to duplicate the mamba cache in case of decoding with n>1 (similar to CacheEngine.copy()) , Lines 981-982 are actually responsible for copying the mamba cache after the forward pass,  , from the `mamba_cache4gc` "buffer"  back into the right place in mamba cache. `mamba_cache4gc` acts as a fixed size memory buffer that his sole purpose is to inhabit the mamba cache during the decoding process. We cannot use the same buffer `mamba_cache` both as the fixed size memory for GC and the persistent MambaCache since it will override existing data. Ideally, It would be best if we could have a "PagedMamba" implementation where we would not need to do those exchanges in the memory.

File: vllm/worker/model_runner.py
Comment (mzusman) [reply vllm/worker/model_runner.py]: I agree that this model-specific code and it doesn't suppose to be in `model_runner`, It's there at the moment so we could access the request information and pick up the correct mamba cache indices before the forward pass. Moving this code as it is to the model implementation along with all of the other mamba specific code means that we will need to access the block tables / slot mapping in some way that is compatible with mamba (Fixed size space).
I was thinking that maybe this specific piece of code could fit into a more abstract version of CacheEngine, where the `CacheEngine` will not be KV-specific and will derive its' cache shape from the model implementation/config.

File: vllm/transformers_utils/configs/jamba.py
Comment (mzusman) [reply vllm/transformers_utils/configs/jamba.py]: Fixed, It is not required anymore after `Jamba` was merged into `transformers`

File: vllm/worker/cache_engine.py
Comment (mzusman) [reply vllm/worker/cache_engine.py]: Fixed

File: vllm/worker/model_runner.py
Comment (zhuohan123) [inline vllm/worker/model_runner.py:870]: This part is for forking sequences and parallel decoding.

File: vllm/worker/model_runner.py
Comment (zhuohan123) [inline vllm/worker/model_runner.py:883]: This part is to contiguous the tensor.

**Comment by zhisbug** [comment]
May I get an update regarding the status of this PR?
It seems the author stopped working on it?

**Comment by mzusman** [comment]
> May I get an update regarding the status of this PR? It seems the author stopped working on it?

We're currently still working on it, The PR works well, we're just waiting for a more suitable cache blocks manger that will be able to accommodate the Mamba cache (seqlen agnostic cache) instead of just managing it directly inside each worker. We're in direct contact with the vLLM team in that regard.

File: vllm/model_executor/input_metadata.py
Comment (mzusman) [reply vllm/model_executor/input_metadata.py:1]: Yes, cleaned it up 👍

File: vllm/core/scheduler.py
Comment (tomeras91) [inline vllm/core/scheduler.py:375]: rename
`flush_last_step_finished_req_ids`

File: vllm/core/scheduler.py
Comment (tomeras91) [inline vllm/core/scheduler.py:1027]: bugfix - don't change relative to main

File: vllm/core/scheduler.py
Comment (tomeras91) [inline vllm/core/scheduler.py:377]: typo - finisehd instead of finished

File: vllm/sequence.py
Comment (tomeras91) [inline vllm/sequence.py:825]: Fix docstring

File: vllm/worker/model_runner.py
Comment (tomeras91) [inline vllm/worker/model_runner.py:170]: remove

File: vllm/worker/model_runner.py
Comment (tomeras91) [inline vllm/worker/model_runner.py:1025]: consistency - change to .update()

File: vllm/worker/worker.py
Comment (tomeras91) [inline vllm/worker/worker.py:246]: align naming

File: vllm/model_executor/models/jamba.py
Comment (tomeras91) [inline vllm/model_executor/models/jamba.py:1]: check vs our internal fork

File: vllm/model_executor/models/jamba.py
Comment (tomeras91) [inline vllm/model_executor/models/jamba.py:676]: rename seqlen_agnostic_cache to mamba_cache

File: vllm/worker/model_runner.py
Comment (tomeras91) [reply vllm/worker/model_runner.py:1025]: rename `get_seqlen_agnostic_capture_inputs`

File: vllm/config.py
Comment (mzusman) [reply vllm/config.py:352]: Yeah , tests failed because they didn't have num_hidden_layers in HF config , see https://github.com/vllm-project/vllm/blob/cd9ba35121abca25bf8a3f9c7cb2bb091746416e/vllm/config.py#L244 they also treat it there this way

File: vllm/model_executor/models/__init__.py
Comment (mzusman) [reply vllm/model_executor/models/__init__.py:64]: I don't see they sort it alphabetically, Putting it last will avoid conflicts

File: vllm/worker/embedding_model_runner.py
Comment (tomeras91) [reply vllm/worker/embedding_model_runner.py:60]: But you didn't add `finished_requests_ids` to `execute_model` of `ModelRunnerBase` (or of any other model runner class..)

File: vllm/worker/cpu_model_runner.py
Comment (mzusman) [reply vllm/worker/cpu_model_runner.py:328]: the formatter doesn't like

File: vllm/worker/embedding_model_runner.py
Comment (mzusman) [reply vllm/worker/embedding_model_runner.py]: You're right, fixed it

File: vllm/worker/cache_engine.py
Comment (mzusman) [reply vllm/worker/cache_engine.py:37]: formatter doesn't let me

File: vllm/worker/embedding_model_runner.py
Comment (tomeras91) [inline vllm/worker/embedding_model_runner.py:56]: Reduce diff from `main` branch here

File: vllm/worker/model_runner.py
Comment (mzusman) [reply vllm/worker/model_runner.py:923]: Agree, but the formatter doesn't let me

File: vllm/core/scheduler.py
Comment (tlrmchlsmth) [inline vllm/core/scheduler.py:294]: Could you expand on the comment a bit more, and define the contract between the scheduler and the model for the `finished_request_ids`? I think that we should be able to say that this lets the model know that any state associated with these requests can and must be released after the current step.

Actually I have a question about this: Is it the case that the model holds onto stale state? E.g. if the model generates the last token for a request at inference step N, does it not free that request until inference step N + 1?

File: vllm/model_executor/models/jamba.py
Comment (tlrmchlsmth) [inline vllm/model_executor/models/jamba.py:785]: Why is this necessary? Could you add a comment?

File: vllm/worker/cache_engine.py
Comment (mzusman) [reply vllm/worker/cache_engine.py]: Yeah, I added a comment and changed the `num_layers` var to `num_attention_layers` to further clarify this idea.

File: tests/models/test_jamba.py
Comment (mzusman) [reply tests/models/test_jamba.py:18]: That's right, I've tested it with float16/bfloat16 and haven't had good results with either. We speculate that it happens because the tiny model weights are randomly generated.

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:294]: Done,
RE Question: That's right, we've done that to avoid sending another signal to workers upon finishing requests. And also have as few changes as possible in the workers/engine code, so we've settled on this solution.

File: vllm/model_executor/models/jamba.py
Comment (mzusman) [reply vllm/model_executor/models/jamba.py]: Yes, done. I've added several comments to the Jamba code to make it more readable.

**Comment by mzusman** [comment]
AFAIU CI distributed-tests-2-gpus test fails regardless of this PR.

File: vllm/config.py
Comment (cadedaniel) [inline vllm/config.py:375]: nit: do we need two methods for this op? if it's just for code cleanliness let's make the other one private (`_get_num_seqlen_agnostic_layers`)

File: vllm/core/scheduler.py
Comment (cadedaniel) [inline vllm/core/scheduler.py:370]: nit: suggest `get_and_reset_finished_request_ids` so that the reader does not need to wonder where it is "flushed" to

File: vllm/core/scheduler.py
Comment (cadedaniel) [inline vllm/core/scheduler.py:1042]: There's a leak here if a sequence is running then is preempted, and aborted by the user while it is preempted. We should check all queues.

File: vllm/worker/model_runner.py
Comment (cadedaniel) [inline vllm/worker/model_runner.py:1041]: nit: can we add a comment explaining why this is here? e.g. it is only used by mamba model cuda graph

File: vllm/config.py
Comment (cadedaniel) [reply vllm/config.py:375]: nit: suggest small docstring explaining an example when this would be true (e.g. Mamba/SSM models)

File: vllm/core/scheduler.py
Comment (cadedaniel) [inline vllm/core/scheduler.py:296]: nit: consider making private `_finished_request_ids` so users don't accidentally access it directly.

Comment (cadedaniel) [reply vllm/core/scheduler.py:296]: ugh I guess the scheduler doesn't do this for anything.. nvm we can clean up later

File: vllm/model_executor/models/jamba.py
Comment (cadedaniel) [inline vllm/model_executor/models/jamba.py:681]: nit: suggest adding comment explaining when this would happen / why it's dynamic

**Review by cadedaniel — APPROVED** [review]
This PR looks good to me. I will push to get it merged (we also want to use the new finished_request_ids bookkeeping in https://github.com/vllm-project/vllm/pull/5765#issuecomment-2196190703).

One edge-case that is missed is in the scheduler, where a certain lineage of a sequence can cause it to be finished without ever notifying the Worker. I would rather get this merged and fix that in another PR given it's so esoteric.

I left some nits; feel free to fix impactful ones now or in a follow up PR. Want to get this merged to avoid more rebase work.

I will get another reviewer for the requirements-mamba and Dockerfile. Other parts LGTM.

**Comment by WoosukKwon** [comment]
QQ: Does this PR support parallel sampling (i.e., `n` > 1 in sampling params)? While I don't think it is not necessary to support parallel sampling in this PR, I'd like to know if this case was considered. Supporting parallel sampling might be a bit hard since it requires implementing `copy_blocks` for the Mamba cache. If this is not trivial, please leave a comment on the code.

**Comment by mzusman** [comment]
> QQ: Does this PR support parallel sampling (i.e., `n` > 1 in sampling params)? While I don't think it is not necessary to support parallel sampling in this PR, I'd like to know if this case was considered. Supporting parallel sampling might be a bit hard since it requires implementing `copy_blocks` for the Mamba cache. If this is not trivial, please leave a comment on the code.

This PR does support parallel sampling, We transfer the seq_ids along with the requests_ids into the Jamba inner state mapping and copy the blocks accordingly, [reference in the Jamba code](https://github.com/vllm-project/vllm/blob/b7e31e3ba5253d16bb4a03bd121e64845b896a38/vllm/model_executor/models/jamba.py#L737-L747)

File: vllm/config.py
Comment (mzusman) [reply vllm/config.py:375]: Yeah, changed `_get_num_seqlen_agnostic_layers` to private and added a small docstring

File: vllm/worker/model_runner.py
Comment (mzusman) [reply vllm/worker/model_runner.py:1041]: Added a comment

File: vllm/model_executor/models/jamba.py
Comment (mzusman) [reply vllm/model_executor/models/jamba.py:681]: Added a comment

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:1042]: I changed it to check through all of the queues, not just the running queue.

File: vllm/core/scheduler.py
Comment (mzusman) [reply vllm/core/scheduler.py:296]: Changed `_finished_request_ids ` to private

**Review by tlrmchlsmth — APPROVED** [review]
The PR looks good to me as well.

**Comment by mzusman** [comment]
Tests failed due to timeouts to HF
Ready to be merged

File: vllm/engine/llm_engine.py
Comment (mmoskal) [reply vllm/engine/llm_engine.py:860]: I have now realized that this code path is not hit it seems, at least in common circumstances due to this https://github.com/vllm-project/vllm/blob/main/vllm/engine/async_llm_engine.py#L563 but probably good to fix anyways!

**Pull request merged**
